plat-rpi3: Merge support for RPi3 to OP-TEEs master branch

jbech-linaro opened the pull request:
This is the patch I received from Philip, I've more or less just applied and rebased it.

Signed-off-by: Joakim Bech <joakim.bech@linaro.org>

jforissier commented on core/arch/arm/plat-rpi3/conf.mk:
```diff
@@ -0,0 +1,74 @@
+$(call force,CFG_ARM32_core,y)
+endif
+
+CFG_NUM_THREADS ?= 8
```
Isn't RPi3 a quad-core CPU? If so, `CFG_NUM_THREADS ?= 4` should be enough.

jbech-linaro replied:
Yes, it is, will fix.

jforissier commented on core/arch/arm/plat-rpi3/conf.mk:
```diff
@@ -0,0 +1,74 @@
+CFG_CRYPTO_WITH_CE ?= n
+CFG_WITH_STACK_CANARIES ?= y
+
+#CFG_TEE_CORE_EMBED_INTERNAL_TESTS ?= n
```
Remove

jforissier commented on core/arch/arm/plat-rpi3/conf.mk:
```diff
@@ -0,0 +1,74 @@
+arm64-platform-cflags += -Wno-error=cast-align
+
+#libutil_with_isoc := y
+#libtomcrypt_with_optimize_size := y
```
Remove also

jbech-linaro commented on core/arch/arm/plat-rpi3/console.c:
```diff
@@ -0,0 +1,118 @@
+/*
```
Code indentation is this file is a bit wonky. I'll have a look at that.

jforissier commented on core/arch/arm/plat-rpi3/conf.mk:
```diff
@@ -0,0 +1,74 @@
+CFG_CRYPTO_SHA256_ARM64_CE := n
+CFG_CRYPTO_SHA1_ARM32_CE := n
+CFG_CRYPTO_SHA1_ARM64_CE := n
+CFG_CRYPTO_AES_ARM64_CE := n
```
All these should be `$(call force,CFG_xxx,n)` so that an error is printed out if the user tries to override the value (e.g. from the command line). You may want to group them with the above ones.

Reminder because it's not obvious... `:=` should not be used to set configuration variables, because it does not play well with environment variables which we want to support properly (some users like to do "CFG_xxx=y make" or "export CFG_xxx=y ; make" rather than "make CFG_xxx=y"). So:
- Use `?=` to set a default value in the global config file (`mk/config.mk`) or in the platform-specific config files like this one,
- Use `$(force ...)` when only a specific value is acceptable (as is the case here).

jbech-linaro replied:
I agree, I will update it.

jbech-linaro commented on core/arch/arm/plat-rpi3/console.c:
```diff
@@ -0,0 +1,118 @@
+  *reg = (unsigned char)value;
+}
+
+# define UART_TX       0
```
I'll move those to the top of the file.

jenswi-linaro replied:
Why not use the existing driver in `core/drivers/serial8250_uart.c`?

jbech-linaro replied:
Yes, that is relevant comment. I have no idea if Phil has another intention or he just didn't think about it.  I'll have a look at it.

jforissier commented on core/arch/arm/plat-rpi3/console.c:
```diff
@@ -0,0 +1,118 @@
+
+int _console_init(unsigned long base_addr)
+{
+  base = (unsigned int *)base_addr;
```
Tabs (I think all occurrences will be picked up by checkpatch)

jbech-linaro commented on core/arch/arm/plat-rpi3/console.c:
```diff
@@ -0,0 +1,118 @@
+int _console_init(unsigned long base_addr)
+{
+  base = (unsigned int *)base_addr;
+  return(0);
```
I don't like the `(0)` thing .. I'll remove parenthesis.

jforissier commented on core/arch/arm/plat-rpi3/console.c:
```diff
@@ -0,0 +1,118 @@
+static unsigned int serial8250_in(int offset)
+{
+  unsigned char *ptr = (unsigned char *) console_base();
+  volatile unsigned int *reg = (unsigned int *)(&ptr[offset]);
```
This file should use `read32()` and `write32()`

jbech-linaro commented on core/arch/arm/plat-rpi3/main.c:
```diff
@@ -0,0 +1,94 @@
+/*
```
Code indentation etc here also .. I'll fix.

jbech-linaro commented on core/arch/arm/plat-rpi3/main.c:
```diff
@@ -0,0 +1,94 @@
+# include <tee/entry_fast.h>
+# include "rpi3_console.h"
+
+static void main_fiq(void);
```
I've remove this and move the implementation up here.

jforissier commented on core/arch/arm/plat-rpi3/console.c:
```diff
@@ -0,0 +1,118 @@
+
+# include <mm/core_memprot.h>
+# include <platform_config.h>
+# include "rpi3_console.h"
```
Why not put the `.h` and `.c` under `core/drivers`?

jbech-linaro commented on core/arch/arm/plat-rpi3/platform_config.h:
```diff
@@ -0,0 +1,132 @@
+#define DRAM0_BASE		0x00000000
+#define DRAM0_SIZE		0x40000000
+
+#define CFG_SHMEM_START		(0x08000000) //below ATF
```
Will change `//` comment.

jforissier commented on core/arch/arm/plat-rpi3/main.c:
```diff
@@ -0,0 +1,94 @@
+
+void console_putc(int ch)
+{
+  _console_putc(ch);
```
That's not a proper name (too generic) especially if the console driver is moved under `core/drivers`. Should be `rpi3_console_putc()` or something like that (what's the UART chip in RPi3?)

jforissier commented on core/arch/arm/plat-rpi3/platform_config.h:
```diff
@@ -0,0 +1,132 @@
+
+# define RPI3_CONFIG
+
+# if 0
```
Remove

jbech-linaro replied:
Will fix

jforissier commented on core/arch/arm/plat-rpi3/platform_config.h:
```diff
@@ -0,0 +1,132 @@
+#ifndef PLATFORM_CONFIG_H
+#define PLATFORM_CONFIG_H
+
+# define RPI3_CONFIG
```
Not used?

jbech-linaro replied:
Can't see that it is used, I'll remove it.

jforissier commented on core/arch/arm/plat-rpi3/platform_config.h:
```diff
@@ -0,0 +1,132 @@
+# endif
+# endif
+
+# define CFG_USE_UBOOT_HOOKS
```
Not used?

jbech-linaro replied:
Agree, I'll remove.

jforissier commented on core/arch/arm/plat-rpi3/platform_config.h:
```diff
@@ -0,0 +1,132 @@
+# undef CFG_TEE_CORE_TA_TRACE
+# endif
+
+# define CFG_DEBUG_VM_ALL 0
```
Either remove or make it a proper configuration variable in `conf.mk` (`CFG_DEBUG_VM_ALL ?= n`). What does it mean BTW?

jbech-linaro replied:
conf.mk, yes. What is is about, I have no idea, we should ask Phil. It looks like a quick hack to enable some more debug prints. I don't think that we should fiddle with trace levels like that. So therefore I lean more towards just remove this for now and let Phil send a patch later on if he needs it. What do you think @jforissier ?

jforissier replied:
Yep, sounds good

jforissier commented on core/arch/arm/plat-rpi3/platform_config.h:
```diff
@@ -0,0 +1,132 @@
+#endif
+#endif /* ARM64 */
+
+/* 16550 UART */
```
The driver contains `serial8250`, here it says 16550. We already have drivers called `core/drivers/ns16550.c` and `core/drivers/serial8250_uart.c`. Could they be re-used/refactored?

jforissier commented on core/arch/arm/plat-rpi3/platform_flags.mk:
```diff
@@ -0,0 +1,61 @@
+ #
+
+# 32-bit flags
+arm32-platform-cpuarch	:= cortex-a15
```
`cortex-a53`?

jbech-linaro replied:
That's wrong, I'll update.

jbech-linaro commented:
I've removed the console.c file now and now we use the `core/drivers/serial8250_uart.c` instead. Regarding the `16550 UART` comment. I'm not sure there, I searched the net a bit, it seems like they use 8250 and 16550 interchangeably. I haven't been able to find out what the actual UART IC in RPi3 is. But with some help with sys fs voodoo from Linus it looks like it indeed is a 16550.
```
root@linaro-developer:/proc# cat /sys/class/tty/ttyS0/type 
3

vim linux/include/uapi/linux/serial.h
...
#define PORT_16550      3
...
```
But, at quite a few pages I see people refer to 8250 when talking about RPi. I'm confused. Anyhow, it works fine with the 8250 driver.

jbech-linaro commented:
And I'll take care of the checkpatch warning as soon as checkpatch completes.

jenswi-linaro commented:
Regarding 16550 vs 8250, see https://en.wikipedia.org/wiki/16550_UART :-)

jbech-linaro commented:
I did look at Wikipedia (amongst others), but I'm not sure if I become much wiser by reading that (more than the "features" indicates a 16550). To me it feels like 16550 is the chip that is used nowadays, but, implementation wise it doesn't matter.

jbech-linaro commented:
I believe travis, report mostly false positives from checkpatch, this is what I get locally
```c
$ git log -3
commit fb5f53fcb478dfcc07b0985bb0b7841231755e71
Author: Joakim Bech <joakim.bech@linaro.org>
Date:   Wed Sep 14 15:49:33 2016 +0200

    [review] Fix checkpatch errors

commit 862adf6b686c12ce492aeeca5174bd675fdca8f2
Author: Joakim Bech <joakim.bech@linaro.org>
Date:   Wed Sep 14 13:09:30 2016 +0200

    fixup! plat-rpi3: Initial support for RPi3

commit 5a5e28e23d0e349b64024cadc0d68243a1d173c3
Author: Philip Attfield <opensource@sequiturlabs.com>
Date:   Sun Jul 10 10:47:44 2016 +0200

    plat-rpi3: Initial support for RPi3
    
    Signed-off-by: Joakim Bech <joakim.bech@linaro.org>
[jbech@manjaro optee_os]
$ git diff bfdfbe74610fc692ddedf45d5f2da520cebccb83 > foo.patch
[jbech@manjaro optee_os]
$ ../linux/scripts/checkpatch.pl --no-tree foo.patch 
WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
#2: 
new file mode 100644

ERROR: Missing Signed-off-by: line(s)

total: 1 errors, 1 warnings, 365 lines checked

foo.patch has style problems, please review.

NOTE: If any of the errors are false positives, please report
      them to the maintainer, see CHECKPATCH in MAINTAINERS.
```
The `Signed-off-by` will be there, when all squashed ...

jforissier commented on core/arch/arm/plat-rpi3/platform_flags.mk:
```diff
@@ -0,0 +1,61 @@
+platform-cflags += -O0
+else
+platform-cflags += -Os
+endif
```
Already taken care of in `core/arch/arm/arm.mk`.

jbech-linaro replied:
I'll fix

jforissier commented on core/arch/arm/plat-rpi3/platform_flags.mk:
```diff
@@ -0,0 +1,61 @@
+endif
+
+platform-cflags += -g
+platform-aflags += -g
```
Same.

jbech-linaro replied:
I'll fix

jforissier commented:
@jbech-linaro when you're checking locally, you're checking *one* commit which is equal to all the commits in the PR, squashed together. Travis, on the other hands, checks every commit in the PR, because in the end we want every one of them to be correct. So when the review is in progress it's expected that we'll get failures although you have fixed the issues, we should just ignore these failures until the final "rebase -i" (with squash). By reviewing which commit fail checkpatch in the Travis log, you should be able to figure things out.
I'm thinking of an improvement: adding a checkpatch call on the diff between the PR branch tip and the target branch (master). That is, more or less what you have run locally. It would help make sure you have addressed all the checkpatch errors.

jbech-linaro commented on core/arch/arm/plat-rpi3/platform_flags.mk:
```diff
@@ -0,0 +1,61 @@
+platform-cflags += -g
+platform-aflags += -g
+
+CFG_ARM32_user_ta := y
```
This is probably nothing that should be here, so I've removed this also.

jenswi-linaro commented on core/arch/arm/plat-rpi3/main.c:
```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2016, Sequitur Labs Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
```
This license is different from the one we're usually using, is this intentional?

jbech-linaro replied:
I don't know the answer whether it was intentional or not. But, it should be on the same format as the rest (which is how the BSD 2-Clause is written). I'll change it.

jenswi-linaro commented on core/arch/arm/plat-rpi3/sub.mk:
```diff
@@ -0,0 +1,33 @@
+ #
```
Please, all this license stuff for two lines that are copied from another file?

jenswi-linaro commented on core/arch/arm/plat-rpi3/main.c:
```diff
@@ -0,0 +1,99 @@
+{
+	serial8250_uart_init(console_base(), CONSOLE_UART_CLK_IN_HZ,
+			     CONSOLE_BAUDRATE);
+	DMSG("\n\n***\n*** OP-TEE on RPi3 ...\n***\n\n");
```
I guess we could do without this debug print.

jbech-linaro replied:
For me it doesn't matter (but since Phil added it, I guess he would like to have it there). Mr maintainer @jforissier , keep it or remove it?

jforissier replied:
I'm leaning towards removing it. The message contains no specific information and we do print an init message already (https://github.com/OP-TEE/optee_os/blob/master/core/arch/arm/kernel/generic_boot.c#L595).

jbech-linaro replied:
Okidoki, I'll remove it.

jbech-linaro commented on core/arch/arm/plat-rpi3/conf.mk:
```diff
@@ -0,0 +1,70 @@
+ #
```
Personally I don't mind, but since it seems to be common practice not only in OP-TEE to not have license information in makefiles etc, I'll remove it.

jenswi-linaro commented on core/arch/arm/plat-rpi3/platform_config.h:
```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2016, Sequitur Labs Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
```
License

jbech-linaro replied:
I've pushed another one.

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented on core/arch/arm/plat-rpi3/conf.mk:
```diff
@@ -0,0 +1,39 @@
+# 32-bit flags
+arm32-platform-cpuarch		:= cortex-a15
```
`cortex-a53`

jbech-linaro replied:
Oh, another one, will fix.

jforissier commented on core/arch/arm/plat-rpi3/platform_flags.mk:
```diff
@@ -0,0 +1,20 @@
+# 32-bit flags
```
This file is not used it seems?

jbech-linaro replied:
What is right and what is wrong here. I can only see `platform_flags.mk` in this commit/folder, but still we refer to it in the `ta/ta.mk`. I.e, this is something used to propagate the platform flags, when building the TA's. But, this isn't something used anylonger? We're exporting it using the `conf.mk` instead. right? I.e, both the `platform_flags.mk` in this PR and the line in `ta/ta.mk` should be removed?

```
$ rgrep platform_flags.mk
ta/ta.mk:       $(wildcard core/arch/$(ARCH)/plat-$(PLATFORM)/platform_flags.mk) \
```

jforissier replied:
I think it's not used since 739804b and `ta/ta.mk` should be updated accordingly (i.e., delete the line you're mentioning). @jenswi-linaro is this correct?

jenswi-linaro replied:
@jforissier, I agree

jbech-linaro replied:
Ok, seems like we're all on the same page here. I'll remove this one and I'll add a separate PR for the `ta/ta.mk` later on.

jbech-linaro commented:
So it seems like we've narrowed it down and seems like it's ready for a squash. I have a RB from Jens. @jforissier , you've been active here also, are you OK with this as it's looking now?

jforissier commented:
@jbech-linaro yes
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jbech-linaro commented:
Patch squashed, rebased and tags applied.

```bash
22561 subtests of which 0 failed
66 test cases of which 0 failed
0 test case was skipped
TEE test application done!
```

```bash
$ ../linux/scripts/checkpatch.pl --no-tree 0001-plat-rpi3-Initial-support-for-RPi3.patch 
WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
#27: 
new file mode 100644

total: 0 errors, 1 warnings, 234 lines checked

0001-plat-rpi3-Initial-support-for-RPi3.patch has style problems, please review.

NOTE: If any of the errors are false positives, please report
      them to the maintainer, see CHECKPATCH in MAINTAINERS.
```
